index.js

```diff
@@ -1,79 +1,45 @@
 'use strict';
 
-const bigquery = require('./lib/bigquery');
+const logger = require('./lib/logger');
 const Inputs = require('./lib/inputs');
 
 exports.handler = (event, context, callback) => {
-  let errs = [];
-  function concatErrors() {
-    errs.forEach(e => exports.logError(`${e}`));
-    if (errs.length === 1) {
-      return errs[0];
-    } else if (errs.length > 1) {
-      let msgs = errs.map(e => `${e}`).join(', ');
-      return new Error(`Multiple errors: ${msgs}`);
-    } else {
-      return null;
-    }
-  }
+  let fatalErr, records;
 
   // decode the base64 kinesis records
   if (!event || !event.Records) {
-    let err = new Error(`Invalid event input: ${JSON.stringify(event)}`);
-    exports.logError(`${err}`);
-    return callback(err);
+    fatalErr = `Invalid event input: ${JSON.stringify(event)}`;
+  } else {
+    records = event.Records.map(r => {
+      try {
+        return JSON.parse(new Buffer(r.kinesis.data, 'base64').toString('utf-8'));
+      } catch (decodeErr) {
+        fatalErr = `Invalid record input: ${JSON.stringify(r)}`;
+      }
+    });
   }
-  let records = event.Records.map(r => {
-    try {
-      return JSON.parse(new Buffer(r.kinesis.data, 'base64').toString('utf-8'));
-    } catch (err) {
-      errs.push(new Error(`Invalid record input: ${JSON.stringify(r)}`));
-      return false;
-    }
-  }).filter(r => r);
 
-  // input records and check for unrecognized
-  let inputs = new Inputs(records);
-  inputs.unrecognized.forEach(r => {
-    errs.push(new Error(`Unrecognized input record: ${JSON.stringify(r)}`));
-  });
+  // NOTE: callback with errors here will cause the lambda to run indefinitely
+  // on the same kinesis records... so you'd better monitor for errors.
+  if (fatalErr) {
+    logger.error(fatalErr);
+    return callback(new Error(fatalErr));
+  }
 
-  // format and insert known inputs
-  let doInserts = inputs.formatRecords().then(tableAndRows => {
-    return Promise.all(tableAndRows.map(data => {
-      return bigquery.insert(data[0], data[1])
-    }));
+  // complain very loudly about unrecognized input records
+  let inputs = new Inputs(records, process.env.PINGBACKS);
+  inputs.unrecognized.forEach(r => {
+    logger.warn(`Unrecognized input record: ${JSON.stringify(r)}`);
```

Comment:
This would be someone (probably DT) writing a record to the kinesis stream that the lambda doesn't know what to do with.  Yeah, should probably be an error.  Will upgrade!